File: 0733-flood-fill/0733-flood-fill.dart
class Solution {
  List<List<int>> floodFill(List<List<int>> image, int sr, int sc, int color) {
    var target = image[sr][sc];
    if (target == color) {
        return image;
    }
    var m = image.length;
    var n = image[0].length;
    var dx = [1,-1,0,0];
    var dy = [0,0,1,-1];
    var visited = List.generate(m, (i) => List.generate(n, (j) => 0));
    visited[sr][sc] = 1;
    var stack = [[sr,sc]];

    while (stack.isNotEmpty) {
        var temp = stack.removeLast();
        var x = temp[0];
        var y = temp[1];
        image[x][y] = color;
        for (var i = 0; i < 4; i ++) {
            var nx = x + dx[i];
            var ny = y + dy[i];
            if (nx < 0 || nx >= m || ny < 0 || ny >= n) {
                continue;
            }
            if (image[nx][ny] == target && visited[nx][ny] ==0) {
                stack.add([nx, ny]);
            }
        }
    }

    return image;
  }
}

File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        dx = [0, 1, 0, -1]
        dy = [1, 0, -1, 0]
        visited = [[0 for _ in range(len(image[0]))] for __ in range(len(image))]
        visited[sr][sc] = 1
        stack = [[sr,sc]]
        value = image[sr][sc]
        if value == color:
            return image
        while stack:
            temp = stack.pop()
            for i in range(4):
                x = temp[0] + dx[i]
                y = temp[1] + dy[i]
                if x >= len(image) or x < 0:
                    continue
                if y >= len(image[0]) or y < 0:
                    continue
                if image[x][y] == value and not visited[x][y]:
                    visited[x][y] = 1
                    stack.append([x, y])
            image[temp[0]][temp[1]] = color
        return image

        